packages/build/src/plugins/child/track.js

```diff
@@ -72,31 +70,41 @@ const forbidMethod = function (keys, method) {
 // `Object.defineProperty(netlifyConfig, key, { value })` or
 // `netlifyConfig.{key} = value`
 // New values are wrapped in a `Proxy` to listen for changes on them as well.
-const trackDefineProperty = function (
-  { parentKeys, configMutations, event },
-  proxy,
-  lastKey,
-  { value, ...descriptor },
-) {
-  const keys = [...parentKeys, lastKey]
+const trackDefineProperty = function ({ parentKeys, configMutations, event }, proxy, key, { value, ...descriptor }) {
+  const keys = [...parentKeys, key]
   const keysString = serializeKeys(keys)
+  const jsonKeys = keys.map(jsonNormalizeKey)
 
+  forbidArrayElementAssign(key, keysString)
   forbidEmptyAssign(value, keysString)
 
   const proxyDescriptor = {
     ...descriptor,
     value: trackObjectMutations(value, keys, { configMutations, event }),
   }
   // eslint-disable-next-line fp/no-mutating-methods
-  configMutations.push({ keys, value, event })
+  configMutations.push({ keys: jsonKeys, value, event })
   logConfigMutation(keysString, value)
-  return Reflect.defineProperty(proxy, lastKey, proxyDescriptor)
+  return Reflect.defineProperty(proxy, key, proxyDescriptor)
 }
 
 const serializeKeys = function (keys) {
   return keys.map(String).join('.')
 }
 
+// `configMutations` is passed to parent process as JSON
+const jsonNormalizeKey = function (key) {
+  return typeof key === 'symbol' ? String(key) : key
+}
+
+const forbidArrayElementAssign = function (key, keyString) {
+  const index = typeof key === 'string' ? Number(key) : key
+  if (Number.isInteger(index)) {
+    throwValidationError(`Setting "netlifyConfig.${keyString}" individual array element is not allowed.
+Please set the full array instead.`)
```

Comment:
In the future, we might want to allow direct array assignments.
It is a little harder to implement, so for the first iteration, we only allow replacing the full array, not mutating individual elements. However, it still possible to change specific array elements by cloning the array, updating the element then replacing the array, so that's ok. This error message ensures plugin authors are aware of this.